aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java

```diff
@@ -0,0 +1,104 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws;
+
+import java.util.Map;
+import java.util.UUID;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.PropertyUtil;
+import software.amazon.awssdk.awscore.client.builder.AwsClientBuilder;
+import software.amazon.awssdk.awscore.client.builder.AwsSyncClientBuilder;
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.glue.GlueClient;
+import software.amazon.awssdk.services.kms.KmsClient;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest;
+
+/**
+ * Example of a {@link AwsClientFactory} for the assume role use case.
+ * <p>
+ * The factory is initialized with a role ARN, region and optional external ID to assume from catalog properties,
+ * and configure all clients except the STS client to use the STS assume role credentials provider.
+ * The STS client is initialized using default credential and region chain
+ * and used to refresh the assume role session token.
+ */
+public class AssumeRoleAwsClientFactory implements AwsClientFactory {
+
+  private static final SdkHttpClient HTTP_CLIENT_DEFAULT = UrlConnectionHttpClient.create();
+
+  private String roleArn;
+  private String externalId;
+  private int timeout;
+  private Region region;
+
+  @Override
+  public S3Client s3() {
+    return S3Client.builder().applyMutation(this::configure).build();
+  }
+
+  @Override
+  public GlueClient glue() {
+    return GlueClient.builder().applyMutation(this::configure).build();
+  }
+
+  @Override
+  public KmsClient kms() {
+    return KmsClient.builder().applyMutation(this::configure).build();
+  }
+
+  @Override
+  public void initialize(Map<String, String> properties) {
+    roleArn = properties.get(AwsProperties.CLIENT_ASSUME_ROLE_ARN);
+    Preconditions.checkNotNull(roleArn,
+        "Cannot initialize AssumeRoleClientConfigFactory with null role ARN");
+    timeout = PropertyUtil.propertyAsInt(properties, AwsProperties.CLIENT_ASSUME_ROLE_TIMEOUT_SEC,
+        AwsProperties.CLIENT_ASSUME_ROLE_TIMEOUT_SEC_DEFAULT);
+    externalId = properties.get(AwsProperties.CLIENT_ASSUME_ROLE_EXTERNAL_ID);
+
+    String regionStr = properties.get(AwsProperties.CLIENT_ASSUME_ROLE_REGION);
+    Preconditions.checkNotNull(regionStr, "Cannot initialize AssumeRoleClientConfigFactory with null region");
+    region = Region.of(regionStr);
+  }
+
+  private <T extends AwsClientBuilder & AwsSyncClientBuilder> T configure(T clientBuilder) {
+    clientBuilder.credentialsProvider(StsAssumeRoleCredentialsProvider.builder()
```

Comment:
As we did in a few other places (like S3OutputStream) it would be good to break out the builders a little as the nested builders tend to be a little harder to follow.